packages/gatsby-transformer-remark/src/extend-node-type.js

```diff
@@ -415,18 +417,20 @@ module.exports = (
       truncate,
       excerptSeparator
     ) {
-      if (excerptSeparator) {
+      // if excerptSeparator in options and excerptSeparator in content then we will get an excerpt from grayMatter that we can use
+      if (excerptSeparator && markdownNode.excerpt !== ``) {
         return markdownNode.excerpt
       }
-      // TODO truncate respecting markdown AST
-      const excerptText = markdownNode.rawMarkdownBody
-      if (!truncate) {
-        return prune(excerptText, pruneLength, `…`)
-      }
-      return _.truncate(excerptText, {
-        length: pruneLength,
-        omission: `…`,
+      const ast = await getMarkdownAST(markdownNode)
+      const excerptAST = await getExcerptAst(ast, markdownNode, {
```

Comment:
In your changes `getExcerptAst` expect first param to be html AST (not markdown AST) (still trying to wrap my head around why this is needed). Is this correct here?

Reply:
@pieh thanks for the review. My understanding is that `getExcerptAst` is a function that will take either kind of AST(html or markdown)

Based on the pruneLength it identifies the last node that should be included. It prunes or truncates that lastNode so that you have an AST that just represents the excerpt. If there is a content separator it will instead use that to generate the tree, and if the pruneLength is less than the length of the markdown it will just return the full tree.

It is being used by getExcerptHtml to get the excerpt when `format: HTML` is specified so I started using it for the markdown case. An alternative would be to simply prune/truncate the rawMarkdownBody (which is what it was doing before: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-remark/src/extend-node-type.js#L422). In that case my change would be simpler and would just be adding a check to this line: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-transformer-remark/src/extend-node-type.js#L418 to check that `markdown.excerpt ! == ""`

I thought that the reason the html method doesn't do this is that it could end up returning strings where the markdown symbols have been truncated ( ie. you could end up with a sentence cut off with only part of some markdown like "Hello **World**" being retuned as "Hello **World*"). I am not sure about this though ( do you know?). I could do some testing to find out.

I will test the above when I get time ( will be next day or so). In the meantime if you have context on it please let me know.

Reply:
@pieh I've added a test that show what I'm talking about ( I hope). The latest test I've added wouldn't pass if we didn't use the AST to generate the pruned excerpt. It would count markdown characters in the pruneLength and would return malformed markdown in some cases. ie. in the test case I've added it would return `Where oh where \*\*is` as opposed to ``Where oh where \*\*is\*\*…\n`

Does this answer your question or where you asking something different?

Reply:
@pieh could you take a look at my reply above? Also, if I misunderstood your question please let me know. Thanks

Reply:
@pieh ^